fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/ObjectAclFactory.java

```diff
@@ -0,0 +1,218 @@
+/**
+ * Copyright 2017 The University of North Carolina at Chapel Hill
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.acl.fcrepo4;
+
+import java.text.ParseException;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.jena.query.QueryExecution;
+import org.apache.jena.query.QuerySolution;
+import org.apache.jena.query.ResultSet;
+import org.apache.jena.rdf.model.RDFNode;
+import org.apache.jena.rdf.model.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import edu.unc.lib.dl.acl.service.PatronAccess;
+import edu.unc.lib.dl.acl.util.UserRole;
+import edu.unc.lib.dl.fedora.PID;
+import edu.unc.lib.dl.rdf.CdrAcl;
+import edu.unc.lib.dl.sparql.SparqlQueryService;
+import edu.unc.lib.dl.util.DateTimeUtil;
+
+/**
+ * Factory which provides access control details which are directly represented
+ * on objects, but does not take into account inheritance.
+ * 
+ * @author bbpennel
+ *
+ */
+public class ObjectAclFactory implements AclFactory {
+
+	private static final Logger log = LoggerFactory.getLogger(ObjectAclFactory.class);
+
+	private LoadingCache<String, List<Entry<String, String>>> objAclCache;
+	private long cacheTimeToLive;
+	private long cacheMaxSize;
+
+	private SparqlQueryService queryService;
+
+	private final List<String> roleUris;
+
+	public ObjectAclFactory() {
+		List<String> roleUris = new ArrayList<>(UserRole.values().length);
+
+		for (UserRole role : UserRole.values()) {
+			roleUris.add(role.getPropertyString());
+		}
+
+		this.roleUris = Collections.unmodifiableList(roleUris);
+	}
+
+	public void init() {
+		objAclCache = CacheBuilder.newBuilder()
+				.maximumSize(cacheMaxSize)
+				.expireAfterWrite(cacheTimeToLive, TimeUnit.MILLISECONDS)
+				.build(new ObjectAclCacheLoader());
+	}
+
+	@Override
+	public Map<String, Set<String>> getPrincipalRoles(PID pid) {
+		String pidString = pid.getRepositoryPath();
```

Comment:
These two lines get called repeatedly in (i.e., 87-88) several of the methods that follow. Would it be worth creating a  private method that returns the objAcls?